sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala

```diff
@@ -137,3 +139,47 @@ trait EvalPythonExec extends UnaryExecNode {
     }
   }
 }
+
+/**
+ * A TaskContext aware iterator.
+ *
+ * As the Python evaluation consumes the parent iterator in a separate thread,
+ * it could consume more data from the parent even after the task ends and the parent is closed.
+ * Thus, we should use ContextAwareIterator to stop consuming after the task ends.
+ */
+class ContextAwareIterator[IN](iter: Iterator[IN], context: TaskContext) extends Iterator[IN] {
+
+  val thread = new AtomicReference[Thread]()
+
+  if (iter.hasNext) {
```

Comment:
Will this change the thread that `iter.hasNext` is running? We can add the listeners without checking it.

Reply:
Actually this is to make sure the upstream iterator is initialized. The upstream iterator must be initialized earlier as it might register another completion listener and the listener should run later than this one.